Implement random teleport command

pop4959 opened the pull request:
Adds `/tpr` and `/settpr` commands, which respectively allow you to teleport randomly or set teleportation parameters.

Server owners are expected to set the center with `/settpr` before players can use `/tpr`. They can also set the minimum and maximum range to be teleported from the center (default 0-1000).

Also includes an event where plugins can adjust or cancel the teleport.

Closes #3154.

pop4959 committed:
Implement random teleport command

JRoy commented on Essentials/src/com/earth2me/essentials/RandomTeleport.java:
```diff
@@ -0,0 +1,55 @@
+
+    public RandomTeleport(final IEssentials essentials) {
+        this.essentials = essentials;
+        File file = new File(essentials.getDataFolder(), "random.yml");
+        config = new EssentialsConf(file);
+        config.setTemplateName("/random.yml");
```
`random` seems like an odd name. Something like `rtp`, `randomtp`, or `randomteleport`

JRoy commented on Essentials/src/com/earth2me/essentials/commands/Commandsettpr.java:
```diff
@@ -0,0 +1,42 @@
+        RandomTeleport randomTeleport = ess.getRandomTeleport();
+        if (args.length == 0 || "center".equalsIgnoreCase(args[0])) {
+            randomTeleport.setCenter(user.getLocation());
+            user.sendMessage(tl("settpr", "center"));
```
English literals ruin the point of translation :P

JRoy commented on Essentials/src/com/earth2me/essentials/RandomTeleport.java:
```diff
@@ -0,0 +1,55 @@
+    public RandomTeleport(final IEssentials essentials) {
+        this.essentials = essentials;
+        File file = new File(essentials.getDataFolder(), "random.yml");
+        config = new EssentialsConf(file);
```
```suggestion
        config = new EssentialsConf(file);
        config.options().header("this is a string stringy xdxdxd");
```

JRoy reviewed (changes requested):
ADSF,JSADFJFDSAASFDASFDAS looks good

pop4959 committed:
rename file

pop4959 committed:
tl fix

pop4959 committed:
improve config

pop4959 committed:
Add biome blocking (no more oceans!) and location finding / caching

pop4959 commented:
Location caching added in the last commit. Included now because with biome blocking, it is possible for a random teleport location to be invalid (if it is the wrong biome). Safe teleportation in Essentials does not handle this (oceans are considered safe). We do not want to have to wait for each location to be checked asynchronously, one at a time, because this may cause a large delay ("lag") in the command. Instead, when a random location is requested, a location will be returned from the cache if possible.

How the caching is handled:

When a random location is requested, it will return one from the cache immediately if possible.

If the number of cached locations is below a threshold, some new locations are asynchronously computed and stored.

If there are no cached locations (should only happen first time the command is run, or if the cache was cleared), then some number of attempts are made to get a random location "on demand" (but still asynchronously). This can take a bit longer.

If there are no cached locations and no "on demand" location is found, the center location is returned as a final resort. This could happen if there is legitimately no valid location to teleport to, or if the cache was empty and you also happen to be very unlucky.

Visual example (running the command several times, printing out the cache size before it does anything else):
![image](https://user-images.githubusercontent.com/17698576/85918156-0f601300-b815-11ea-81c1-a4e31d59a949.png)

pop4959 committed:
optimisation - makes command significantly faster on spigot

pop4959 commented:
Latest commit should prevent caching from blocking the command on spigot, which results in a significant performance boost.

pop4959 committed:
Merge branch '2.x' into command-tpr

pop4959 committed:
Default to worldborder center if none set

pop4959 commented:
Apparently the vanilla World Border feature was added in 1.8, so we can actually use that to set a default automatically. Hooray! 🎉

SrBedrock commented:
Would it be possible to define the minimum and maximum distance per* world?

pop4959 commented:
> Would it be possible to define the minimum and maximum distance according to the world?

It can be set in the config or with the `/settpr` command. It's not automatically set to the world border because 1000 is a more sane default than the 30 million or whatever it normally is.

pop4959 commented:
Did some testing on Spigot and Paper. Results are basically as expected; performance on Spigot fluctuates a bit, while Paper is consistently better because of asynchronous chunk loading.

**Spigot:**

![2020-06-27_22 25 57](https://user-images.githubusercontent.com/17698576/85938808-41c94900-b8c5-11ea-9ba5-fbd1b4f0ae75.png)

**Paper:**

![2020-06-27_20 51 10](https://user-images.githubusercontent.com/17698576/85938657-142fd000-b8c4-11ea-9163-45a97426c7c0.png)

JRoy commented on Essentials/src/com/earth2me/essentials/commands/Commandtpr.java:
```diff
@@ -0,0 +1,124 @@
+        if (event.isCancelled()) {
+            return;
+        }
+        Location center = event.getCenter();
+        double minRange = event.getMinRange();
+        double maxRange = event.getMaxRange();
```
would rather this be inlined

JRoy commented on Essentials/src/com/earth2me/essentials/commands/Commandtpr.java:
```diff
@@ -0,0 +1,124 @@
+        // Try to build up the cache if it is below the threshold
+        if (cachedLocations.size() < randomTeleport.getCacheThreshold()) {
+            ess.getServer().getScheduler().scheduleSyncDelayedTask(ess, () -> {
+                for (int i = 0; i < findAttempts; ++i) {
```
i n l i n e

pop4959 replied:
please elaborate

JRoy reviewed (changes requested):
m i n o r  n i t p i c k s  b u t  e v e r y t h i n g  e l s e  i s  f i n e

pop4959 committed:
i n l i n e

pop4959 committed:
More sensible ddefault max range

pop4959 commented:
Was thinking on this more, and the max range of 1000 is might be a bit too arbitrary/confusing. In addition, if the world border was set to less than 1000 blocks out from spawn this would potentially kill the player. A better and safer default is to just use the world border and let the server admin set it lower if they wish to.

pop4959 committed:
Refactor + add pre-caching

pop4959 commented:
Moved most of the actual random teleportation code into the RandomTeleport class, and decoupled caching logic from the random teleportation function so that it can be called on its own. Added a new `pre-cache` configuration option. This allows locations to already be available to the tpr command when it is run for the first time, making it faster.

mdcfe commented on Essentials/src/tpr.yml:
```diff
@@ -0,0 +1,16 @@
+# Configuration for the random teleport command.
+# Some settings may be defaulted, and can be changed via the /settpr command in-game.
```
Shouldn't this file include all the available config options (ie pre-cache, max-range)?

pop4959 replied:
Arguably yes, although once something gets saved into the file, it is effectively permanent. This can be problematic, so unless the server owner explicitly sets some of these settings. Here's an explanation for the thought process behind putting all of the current settings where they are:

`center`: We cannot have a default in the config, because we cannot assume what the world name is. So instead, when `getCenter` is called, we try reading the config first, but if that isn't available (or the world is no longer valid), we write the config with the center of the current main world, which is an appropriate default. This can further be changed via explicitly running the `/settpr` command or entering manual config.

`min-range`: In the config by default, and also automatically zero if none specified. Zero is a fine default, and the server owner can increase it if they do not want players teleporting into a spawn protected zone, etc.

`max-range`: Not explicitly in the config, because we want the default to be the max range of whatever world the command is being used in. This allows a true random location in the world, when server owners do not specify something lower. We don't want to hard code 3E7 (the default vanilla world border size) into here, because the server owner may have already changed it. It can be lowered if servers don't want a true random location, but rather something closer to spawn (like 1k, 10k, etc).

`excluded-biomes`: Which biomes should be avoided when selecting a random location. Safe teleport doesn't handle this, and will happily randomly teleport players into oceans for example (and fairly often actually) if not otherwise checked. This is set in the config by default since there is something sensible to set it to, and the server owner can adjust it if they want to add/remove anything.

`find-attempts`: Used internally for determining how many times to try finding random locations. This is something that generally should not be changed (higher = potentially more lag, lower = potentially not finding locations). Server owners can adjust it to tweak performance vs optimal behavior, but it's better not written in the config by default in case Essentials wants to adjust the default value in the future. #EasterEgg 

`cache-threshold`: Also used internally to decide when the cache is getting too empty, and needs to be refilled. Not really much reason to need to change this either, in most cases. #EasterEgg

`pre-cache`: Having this on by default makes the command run much faster (particularly, on the first execution, where there may otherwise be no cache). Having a small number of locations cached isn't particularly heavy, so there is not much reason for most server owners to touch this. It still exists as an option to disable though. #EasterEgg

mdcfe committed:
Merge branch '2.x' into command-tpr
